MQTT - Serial bridge

danielztolnai opened the pull request:
Software serial - MQTT bridge implemented as an xdrv. Useful when interfacing with devices having only a serial port for communication like for example projectors and older AV receivers.

Quick info and usage:
 - Configure pins "SerBr Rx" and "SerBr Tx"
 - Set the baud rate through cmnd/sonoff/SBrBaud (Can be anything from 1200 to 115200)
 - Send data through cmnd/sonoff/serialbr
 - Received data is published to stat/sonoff/SERIALBR
 - Delimiter functionality can be enabled by sending the ASCII code of the delimiter to cmnd/sonoff/SBrDelim
 - Delimiter functionality can be disabled by sending 0xff to cmnd/sonoff/SBrDelim
 - Added support for whitespace characters. It is possible to transmit for example a newline character by putting its escaped version ("\n") in the MQTT message

danielztolnai committed:
TasmotaSeria: Support higher baud rates up to 115200

danielztolnai committed:
Fix baud rate setting read

danielztolnai committed:
Make baud rate command self-explanatory

danielztolnai committed:
Implement Serial-MQTT bridge

danielztolnai committed:
Make baud rate a setting

danielztolnai committed:
Add label for all languages

danielztolnai committed:
Store baud rate on single byte, implement trivial setting command

danielztolnai committed:
Reinitialize serial port with new baud rate on change

arendst commented:
You're 8 hours too late. I have just released this functionality. ;-)

danielztolnai commented:
@arendst Your implementation is using the hardware serial. In a software serial implementation you can choose pins, and you will not have the ESP's reset messages on the serial line. Correct me if I've missed something or said something wrong.

arendst commented:
You are right but what currently uses tasmotaserial I see exceptions randomly so I choose the hardware option this time. I'll look into your implementation too.

danielztolnai committed:
Implement delimiter functionality. Use less memory by default

danielztolnai committed:
Use decimal notation for delimiter to allow special characters

danielztolnai committed:
Fix transmission when delimiter is compiled but disabled (0xff)

danielztolnai committed:
Implement transmission of whitespace characters

Whitespace characters can be sent by escaping them in the MQTT message.

This is needed for the reason that the message handler framework drops whitespace characters.

danielztolnai commented:
@arendst I've been testing this (using 38400 bps) for around 48 hours now and it seems to work (for now). I hope it will work for you as well. :-) Above the advantages I have mentioned before, this also has the comfort of working on devices that use the hardware serial for something else, like the older Sonoff Dual.

arendst commented:
Convinced! I will add to next pre-release. Well done.

arendst commented:
Implemented.

danielztolnai commented:
@arendst Please excuse my bluntness, but the implementation in my pull request is still superior to yours.

1. You've disabled sending raw binary data. MQTT itself supports this, and it works well in my implementation.
2. You've also disabled using any binary byte as the delimiter.
3. You've wrapped the data received from the serial port in JSON. I understand that all Tasmota telemetry and status data uses JSON, but in the case of a serial port, it brings more problems to the table than advantages.
4. The code that you use for MQTT input preprocessing removes all whitespace characters using the isspace() function but you have disabled the "\s" escaped space character in the Unescape() function. This makes it impossible to send a space character through the serial port.
5. My implementation has the ability to disable unneeded parts of the code to save flash and memory space.

**Points 1., 2. and 3. make it impossible to use your implementation as a transparent serial bridge**, which is required in multiple serial port use cases. (Where you have to transfer raw binary data)

I am not pushing you to use my implementation as it is, but please fix these issues (1. to 4.) as having a transparent serial bridge has more opportunities than having a text only one. (Also, the transparent implementation has no drawbacks compared to the text only one.)

aidanmountford commented:
Thanks guys.

I was literally looking at writing this last week ...  I will run it up over the weekend and see how I go...   

A
